[MM-53428] Delete empty drafts on upsert

devinbinnie opened the pull request:
#### Summary
When updating a draft, or clearing a draft by removing the message, our app was only updating the message to be blank, and not setting a `DeleteAt` time. This would cause the API to return hundreds of empty drafts that would never show up on the front-end, but would be potentially costly to return.

This PR adds a check to the `UpsertDraft` call to ensure that we set a `DeleteAt` timestamp if the draft is empty, denoting that the user has effectively deleted it. 

I've also added a migration script to clean up any existing drafts by setting their `DeleteAt` time if they are empty.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-53428

-->
```release-note
Fixed an issue where the API for drafts would return empty drafts
```

mattermost-build commented:
E2E tests not automatically triggered, because the PR is not in a mergeable state. Please update the branch with the base branch and resolve outstanding conflicts.

lieut-data commented on server/channels/app/draft.go:
```diff
@@ -61,6 +61,10 @@ func (a *App) UpsertDraft(c *request.Context, draft *model.Draft, connectionID s
 		return nil, model.NewAppError("CreateDraft", "app.draft.save.app_error", nil, nErr.Error(), http.StatusInternalServerError)
 	}
 
+	if draft.Message == "" {
+		a.Srv().Store().Draft().Delete(draft.UserId, draft.ChannelId, draft.RootId)
+	}
+
```
Should we check this before the `Upsert` above?

devinbinnie replied:
I'm not sure if it matters, because if the above query fails (maybe due to locking) we wouldn't want to try the delete. I assume we just want to make sure that we don't save empty messages in the DB without also having marked them deleted.

harshilsharma63 replied:
I'd say in case of deleting, only delete, not upsert and delete (one vs 2 DB calls). Doesn't feel right upserting only to delete it later.

devinbinnie replied:
@harshilsharma63 Wouldn't that mean that there would still be a message in the draft on the DB side? Not sure if that would matter or not.

devinbinnie replied:
@harshilsharma63 Wouldn't that mean that there would still be a message in the draft on the DB side? Not sure if that would matter or not.

lieut-data commented on server/channels/db/migrations/mysql/000110_delete_empty_drafts.up.sql:
```diff
@@ -0,0 +1,2 @@
+UPDATE Drafts
+SET DeleteAt = UpdateAt WHERE Message = ''
```
Out of curiosity, why do we keep rows at all when deleting a draft? What's the use of a soft deleted empty message?

devinbinnie replied:
This I'm not sure, I assume it could be for an audit trail? We don't have an actual `DELETE` function in the SQL store file, so I didn't want to add one and go too far outside the current spec.

lieut-data replied:
I'm wondering if we need to clarify the spec here. As it stands, this table has a theoretical max size equivalent to  `ChannelMembers` if we aren't cleaning up drafts (imagine every user typing then undoing in all their channels, dms, etc. exactly once over an organic period of time [like me, apparently ;P]). It feels like maybe just an oversight in the original plan?

harshilsharma63 replied:
Maybe include drafts in data-retention? But anyways, beyond the scope of this PR.

devinbinnie replied:
Yeah I'd like to keep the scope to at least just making sure the drafts aren't leaking through the API. I think @lieut-data does make a good point that it's worth investigating if we need to keep the rows or not.

lieut-data replied:
I'd like to pushback on keeping the scope to the the first half of the problem -- we're already having to do a data migration here, and it would be a shame to do another one to delete the empty rows just because we split the work in two artificially.

devinbinnie replied:
Okay, let's have a discussion in the ICU channel about this. We'll get everyone's opinion involved and see what the best direction is :)

agnivade replied:
I agree with Jesse here. There is no value in preserving empty rows in the DB that I can see.

devinbinnie replied:
Doesn't seem like anyone else disagrees so I will move to actually deleting the empty rows.

devinbinnie replied:
@lieut-data Made changes if you'd like to look again

lieut-data commented on server/channels/db/migrations/postgres/000110_delete_empty_drafts.up.sql:
```diff
@@ -0,0 +1,2 @@
+UPDATE drafts
+SET deleteat = updateat WHERE message = ''
```
@agnivade, is this potentially a case where a mass UPDATE (or DELETE) could cause locking issues?

agnivade replied:
Since there is a filter here, that should be fine. What we should avoid is running `UPDATE` on the entire table, for example, when adding a new column that needs to be filled with data. 

Unless potentially _all_ rows could have message as empty? I think that would be very improbable.

devinbinnie replied:
@agnivade It's definitely unlikely but could be possible if everyone just deleted all their drafts at the same time :P

agnivade replied:
Can you clarify what situations lead to an empty message field? Is it only when users delete their drafts? Or are there others?

devinbinnie replied:
I think that's the only situation I can think of

agnivade replied:
Let me run some queries on Community to verify the state.

agnivade replied:
Ok, so for community 83% of the rows are empty. That's a whole lot IMO, so I would be leaning towards making this a batch migration. This will result in the migration becoming a lot more complicated now, but I don't see any other way if we are to uphold our "no locking migrations" objective.

lieut-data reviewed (changes requested):
Thanks for tracking this one down, @devinbinnie! A few clarifying questions below.

agnivade reviewed (commented):
@devinbinnie - We need to calculate the time for these SQL queries to run, along with the size and spec of the DB instances so that they can be part of the release notes.

harshilsharma63 commented on server/channels/app/draft.go:
```diff
@@ -61,6 +61,10 @@ func (a *App) UpsertDraft(c *request.Context, draft *model.Draft, connectionID s
 	}
 
+	if draft.Message == "" {
+		a.Srv().Store().Draft().Delete(draft.UserId, draft.ChannelId, draft.RootId)
```
Missing error handling.

agnivade commented:
@devinbinnie - This still doesn't change the fact that the migration will be locking. Sorry if that was misunderstood somehow. We'll need to do a batch migration anyways.

devinbinnie commented:
> @devinbinnie - This still doesn't change the fact that the migration will be locking. Sorry if that was misunderstood somehow. We'll need to do a batch migration anyways.

@agnivade Yes I know, I just wanted to highlight the changes done to fully remove the draft rows going forward. We will proceed with adding the batch migration once we have a good example of it :)

agnivade commented:
Perfect, sounds good.

lieut-data reviewed (changes requested):
(Blocking review until we sort out the batch migration strategy.)

devinbinnie commented:
@agnivade @lieut-data @harshilsharma63 I just got back from vacation, are we still waiting on a batch migration strategy? Does it make sense to start a thread about this somewhere on community?

agnivade commented:
Yes, that hasn't changed. Please open a thread on Community. cc Ibrahim and me.

agnivade commented:
Feel free to request a review when it's ready.

devinbinnie commented on server/channels/store/storetest/draft_store.go:
```diff
@@ -275,24 +275,6 @@ func testGetDraft(t *testing.T, ss store.Store) {
 		assert.Equal(t, draft2.ChannelId, draftResp.ChannelId)
 	})
-
-	t.Run("get draft including deleted", func(t *testing.T) {
```
Ended up deleting this test since we are hard deleting drafts from the DB. Is there any further cleanup we should do here? We can remove the `includeDeleted` field, and maybe even the `DeleteAt` column altogether?

harshilsharma63 reviewed (approved):
👌👌👌

devinbinnie commented on server/channels/jobs/delete_empty_drafts_migration/delete_empty_drafts_migration.go:
```diff
@@ -0,0 +1,205 @@
+			return
+		case <-time.After(timeBetweenBatches):
+			// Check if there is metadata for that job.
+			// If there isn't, it will be empty by default, which is the right value.
+			userID := job.Data["user_id"]
+			createAt, appErr := worker.getJobMetadata(job, "create_at")
+			if appErr != nil {
+				mlog.Error("DeleteEmptyDraftsMigrationWorker: failed to get create at", mlog.String("worker", worker.name), mlog.String("job_id", job.Id), mlog.Err(appErr))
+				worker.setJobError(job, appErr)
+				return
+			}
```
The call for `SetJobProgress` also wants an int to indicate how far along the job is.

Is this something that's required or should I just put 0? Not sure if there's a good performant way of actually tracking the job progress correctly here.

devinbinnie replied:
Oh nevermind, read the below comment :D

devinbinnie commented on server/channels/app/draft.go:
```diff
@@ -56,6 +56,15 @@ func (a *App) UpsertDraft(c *request.Context, draft *model.Draft, connectionID s
 		return nil, model.NewAppError("CreateDraft", "app.user.get.app_error", nil, nErr.Error(), http.StatusInternalServerError)
 	}
 
+	// If the draft is empty, just delete it
+	if draft.Message == "" {
```
Is that something we support for server-side synced drafts?

larkox replied:
I would expect to, but I don't know really :sweat_smile:

agnivade commented on server/channels/store/sqlstore/draft_store.go:
```diff
@@ -236,3 +233,78 @@ func (s *SqlDraftStore) determineMaxDraftSize() int {
+			)
+			AND d.Message = ''
+		`
+	} else {
+		mlog.Warn("No implementation found to determine the maximum supported draft size")
+	}
```
No need for this check. MM config is always validated to be one of those two drivers during startup.

agnivade reviewed (commented):
Nearly there Devin. Thanks for your patience!

devinbinnie commented on server/channels/jobs/delete_empty_drafts_migration/delete_empty_drafts_migration.go:
```diff
@@ -0,0 +1,206 @@
+	} else if !claimed {
+		return
+	}
+
```
I added it underneath that and set it to Pending. Not sure if there's another step?

agnivade commented on server/channels/jobs/delete_empty_drafts_migration/delete_empty_drafts_migration.go:
```diff
@@ -0,0 +1,224 @@
+
+	// Wait for all clusters to finish DB migration
+	clusterStatus := worker.app.GetClusterStatus()
+	if len(clusterStatus) > 0 {
```
Maybe more efficient to check for `length > 1` ?

agnivade commented on server/channels/jobs/delete_empty_drafts_migration/delete_empty_drafts_migration.go:
```diff
@@ -0,0 +1,224 @@
+		for i := 1; i < len(clusterStatus); i++ {
+			if clusterStatus[i].SchemaVersion != clusterStatus[0].SchemaVersion {
+				// Just wait for the next loop
+				worker.jobServer.SetJobPending(job)
```
Nice, this is much more elegant.

agnivade reviewed (commented):
Looks really good! I will run some final tests tomorrow my time and approve this. But no further comments from my side.

agnivade reviewed (approved):
Tested and looks all good. Thanks @devinbinnie !

devinbinnie commented:
@lieut-data Did you want to give this another pass?

lieut-data commented:
I'll take a look today -- thank you, @devinbinnie!

lieut-data commented on server/channels/jobs/delete_empty_drafts_migration/delete_empty_drafts_migration.go:
```diff
@@ -0,0 +1,224 @@
+		return
+	}
+
+	// Wait for all clusters to finish DB migration
```
Do we need to handle the possibility that an older node starts after the job and starts creating new draft entries? Or even if the cluster is downgraded to recover from an issue and later re-upgraded?

agnivade commented on server/channels/store/sqlstore/draft_store.go:
```diff
@@ -236,3 +233,72 @@ func (s *SqlDraftStore) determineMaxDraftSize() int {
+				d.UserId = dd.UserId 
+				AND d.ChannelId = dd.ChannelId 
+				AND d.RootId = dd.RootId
+				AND d.Message = ''
```
The problem is to store the offsets. You could do everything in one query, but then MySQL does not support `RETURNING` in the DELETE query. So you need one query to calculate the offsets, another to DELETE.

lieut-data replied:
Do we care about the offsets? Is that just used to report job metadata?

agnivade replied:
We care about offsets to survive server restarts. It is an important aspect to make the job resumable. You can restart the server in the middle of a migration, and it picks right off from where it left.

lieut-data replied:
Apologies that I'm still missing something, @agnivade. Here's a more concrete pseudo-code example of my suggestion:

```go
for {
    numRowsDeleted, err := sql.Execute("DELETE FROM Drafts d WHERE d.Message = '' LIMIT 100")
    if numRowsDeleted == 0 {
        break
    }

    time.Sleep(1 * time.Second)
}
```

This pseudo-code will eventually delete all messages with empty drafts in chunks of 100 at a time. It doesn't care about offsets, because it doesn't care about which messages it needs to delete: just that they all go away. It doesn't matter if the code gets restarted: if there are any messages that still need to be deleted, it will find them and pick up from there.

One downside to this approach is presumably the need to scan the table for `d.Message = ''` in the first place. But given we're deleting the rows behind us, and we already know there are a disproportionate number of empty messages compared to "real drafts", we need to look at most of the rows anyway.

Thoughts?

devinbinnie replied:
I'm wondering if the offsets are to avoid locking the whole table every time while the job is running? Not sure if the table is partitioned or not (could be under the hood) but the offsets make sure that we're only working with a specific data set, so the DB won't lock the whole table.

It also means that we can leverage the indexes so that we can grab the rows we want to work with quickly, then we only scan 100 rows at a time maximum while checking for blanks, whereas we'd be scanning more without the offsets.

devinbinnie replied:
A simpler reason for Postgres:
```
ERROR:  syntax error at or near "LIMIT"
LINE 1: DELETE FROM Drafts d where d.Message = '' LIMIT 100
```

I'm guessing that batching is the only way for Postgres?

agnivade replied:
@lieut-data - `LIMIT` on DELETEs is only supported for MySQL unfortunately. :(

But even if we hadn't supported Postgres, my main objective here was to create an example migration that can serve as a template for future ones. I think overall, this is a pretty simple and straightforward migration which shouldn't take too long.

devinbinnie commented on server/channels/jobs/delete_empty_drafts_migration/delete_empty_drafts_migration.go:
```diff
@@ -0,0 +1,224 @@
+
+	// Wait for all clusters to finish DB migration
+	clusterStatus := worker.app.GetClusterStatus()
+	if len(clusterStatus) > 1 {
```
Lol whoops will remove :)

devinbinnie commented on server/channels/store/sqlstore/draft_store.go:
```diff
@@ -236,3 +233,72 @@ func (s *SqlDraftStore) determineMaxDraftSize() int {
+				sq.Gt{"UserId": userId},
+			},
+		}).
+		OrderBy("CreateAt", "UserId ASC").
```
Does it matter to have `ASC` for either actually? Been a while since I've played in SQL but it should sort strings alphabetically and integers in numerical order by default right?

lieut-data replied:
Yeah, I'd be fine with removing it altogether :)

devinbinnie commented on server/channels/store/sqlstore/draft_store.go:
```diff
@@ -236,3 +233,72 @@ func (s *SqlDraftStore) determineMaxDraftSize() int {
+		`
+	}
+
+	if _, err := s.GetMasterX().Exec(sql, createAt, createAt, userId); err != nil {
```
@lieut-data Have a look when you have a second :)

yasserfaraazkhan commented:
/e2e-test

mattermost-build commented:
Successfully triggered E2E testing!
[GitLab pipeline](https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/368140) | [Test dashboard](https://automation-dashboard.vercel.app/cycle/368140)

yasserfaraazkhan reviewed (approved):
verified that the response of API 'api/v4/users/me' match with the Frontend draft count seen

lieut-data reviewed (changes requested):
Thanks, @devinbinnie -- just a few more thoughts as I worked through the specifics of the migration code.

@agnivade, I'm worried about this job pattern. It makes sense to use this pattern to execute the migration, but the actual code to delete these empty drafts is chock full of copy and pasted nuances from other jobs. I can't meaningfully reason about the "job correctness."

I'm wondering if we should take the time -- before merging this PR -- to actual abstract a `MigrationJobWorker` or some such that's provably correct, and maybe even unit tested. Thoughts?

devinbinnie commented on server/channels/jobs/delete_empty_drafts_migration/delete_empty_drafts_migration.go:
```diff
@@ -0,0 +1,222 @@
+
+	// Check if there is metadata for that job.
+	// If there isn't, it will be empty by default, which is the right value.
+	userID := job.Data["user_id"]
+	createAt, appErr := worker.getJobMetadata(job, "create_at")
```
`getJobMetadata` is just a helper function for converting the data to a `string`, it's a bit misnamed.
I think for this job I can remove it since we're just using it in the one place, and it is confusing.

devinbinnie commented on server/channels/jobs/delete_empty_drafts_migration/delete_empty_drafts_migration.go:
```diff
@@ -0,0 +1,222 @@
+
+	// Note that if this fails, then the job would have still succeeded.
+	// So it will try to run the same job again next time, but then
+	// it will just fall through everything because all files would have
+	// converted. The actual job is idempotent, so there won't be a problem.
```
Must have missed that one.

devinbinnie commented on server/channels/jobs/delete_empty_drafts_migration/delete_empty_drafts_migration.go:
```diff
@@ -0,0 +1,222 @@
+				return
+			}
+
+			err = worker.store.Draft().DeleteEmptyDraftsByCreateAtAndUserId(createAt, userID)
+			if err != nil {
+				mlog.Error("DeleteEmptyDraftsMigrationWorker: Failed to delete the empty drafts for the page. Exiting", mlog.Err(err))
+				worker.setJobError(job, model.NewAppError("DoJob", model.NoTranslation, nil, "", http.StatusInternalServerError).Wrap(err))
+				return
+			}
```
Good catch, will fix.

devinbinnie commented:
@lieut-data I've addressed the above feedback (though GitHub is apparently taking time to load the commits). As per your above comment, I think we've already pushed the scope of this change far beyond what was originally intended, and I've spent a lot of time on it already.

I fundamentally agree with you and I like the idea of having some abstraction if we are planning to do future migrations in a similar manner, but I don't feel we should be blocking this change for that, and we can still make those changes in a follow-up change.

agnivade commented:
First of all, I want to say that our team owns schema migrations, and I am to blame for not preparing a sample migration myself. I just prepared the google doc with some pseudo code, and not real code because there was no migration to do until this.

Devin has gone above and beyond to do something as complicated and nuanced as this, and we should not be asking him to do any more. It is also one of the reasons why I didn't ask for unit tests here.

But I personally have tested this PR with several scenarios and verified that it works. So there's some amount of confidence from my side.

> @agnivade, I'm worried about this job pattern. It makes sense to use this pattern to execute the migration, but the actual code to delete these empty drafts is chock full of copy and pasted nuances from other jobs. I can't meaningfully reason about the "job correctness."

Maybe it's just me, and I have been writing a lot of jobs, so I know what works and why does it work.

Anyways, if you want to take over, I have no objections. As long as there is no rush to merge the PR, then it should be fine. Thanks Jesse.

devinbinnie commented:
@lieut-data Definitely happy to hand off the abstraction bits to you Jesse if you have the cycles to work on it :)
I don't believe there's any rush to merge this, but I will need to move on. Happy to answer questions as needed :)

lieut-data commented:
@devinbinnie & @agnivade, in the course of writing unit tests, I discovered a subtle difference between MySQL and Postgres when it comes to deleting empty drafts. 

Imagine a `Drafts` dataset that looks like the following -- essentially 50 drafts with messages, then 50 empty drafts, then another 50 drafts with messages, and another 50 empty drafts:

```
"message", // row 1
"message", // row 2
..
"message", // row 50
"", // row 51
"", // row 52
..
"", // row 100
"message", // row 101
"message", // row 102
..
"message", // row 150
"", // row 151
"", // row 152
..
"", // row 200
```

In each call to `DeleteEmptyDraftsByCreateAtAndUserId` with Postgres, we take the first page of 100 drafts and delete the 50 empty messages within that page. But with MySQL, we start at that window and delete the next 100 empty messages -- in this case, all the empty messages in the table. This is because `DELETE ... LIMIT 100` only puts a limit on how many records are deleted, not the number of records over which we iterate.

In practice, this still works, in that all the empty drafts are deleted. But I'd like your help reasoning about the impact to table locking. Based on https://dev.mysql.com/doc/refman/8.0/en/innodb-locks-set.html:

> For [locking reads](https://dev.mysql.com/doc/refman/8.0/en/glossary.html#glos_locking_read) ([SELECT](https://dev.mysql.com/doc/refman/8.0/en/select.html) with FOR UPDATE or FOR SHARE), [UPDATE](https://dev.mysql.com/doc/refman/8.0/en/update.html), and [DELETE](https://dev.mysql.com/doc/refman/8.0/en/delete.html) statements, the locks that are taken depend on whether the statement uses a unique index with a unique search condition or a range-type search condition.
>  *  For a unique index with a unique search condition, InnoDB locks only the index record found, not the [gap](https://dev.mysql.com/doc/refman/8.0/en/glossary.html#glos_gap) before it.
>  *  For other search conditions, and for non-unique indexes, InnoDB locks the index range scanned, using [gap locks](https://dev.mysql.com/doc/refman/8.0/en/glossary.html#glos_gap_lock) or [next-key locks](https://dev.mysql.com/doc/refman/8.0/en/glossary.html#glos_next_key_lock) to block insertions by other sessions into the gaps covered by the range. For information about gap locks and next-key locks, see [Section 15.7.1, “InnoDB Locking”](https://dev.mysql.com/doc/refman/8.0/en/innodb-locking.html).

I believe we fall into the latter bucket -- does that mean we lock the entire range scanned? In the worst case (exactly 100 deleted messages in the table), it would seem we'd lock the entire table by virtue of the `LIMIT 100` semantics.

Should I rework this to only delete empty messages explicitly in the page in question, like Postgres does?

devinbinnie commented:
> @devinbinnie & @agnivade, in the course of writing unit tests, I discovered a subtle difference between MySQL and Postgres when it comes to deleting empty drafts.
> 
> Imagine a `Drafts` dataset that looks like the following -- essentially 50 drafts with messages, then 50 empty drafts, then another 50 drafts with messages, and another 50 empty drafts:
> 
> ```
> "message", // row 1
> "message", // row 2
> ..
> "message", // row 50
> "", // row 51
> "", // row 52
> ..
> "", // row 100
> "message", // row 101, start of page 2
> "message", // row 102
> ..
> "message", // row 150
> "", // row 151
> "", // row 152
> ..
> "", // row 200
> ```
> 
> In each call to `DeleteEmptyDraftsByCreateAtAndUserId` with Postgres, we take the first page of 100 drafts and delete the 50 empty messages within that page. But with MySQL, we start at that window and delete the next 100 empty drafts -- in this case, all the empty drafts in the table. This is because `DELETE ... LIMIT 100` only puts a limit on how many records are deleted, not the number of records over which we iterate.
> 
> In practice, this still works, in that all the empty drafts are deleted. But I'd like your help reasoning about the impact to table locking. Based on https://dev.mysql.com/doc/refman/8.0/en/innodb-locks-set.html:
> 
> > For [locking reads](https://dev.mysql.com/doc/refman/8.0/en/glossary.html#glos_locking_read) ([SELECT](https://dev.mysql.com/doc/refman/8.0/en/select.html) with FOR UPDATE or FOR SHARE), [UPDATE](https://dev.mysql.com/doc/refman/8.0/en/update.html), and [DELETE](https://dev.mysql.com/doc/refman/8.0/en/delete.html) statements, the locks that are taken depend on whether the statement uses a unique index with a unique search condition or a range-type search condition.
> > 
> > * For a unique index with a unique search condition, InnoDB locks only the index record found, not the [gap](https://dev.mysql.com/doc/refman/8.0/en/glossary.html#glos_gap) before it.
> > * For other search conditions, and for non-unique indexes, InnoDB locks the index range scanned, using [gap locks](https://dev.mysql.com/doc/refman/8.0/en/glossary.html#glos_gap_lock) or [next-key locks](https://dev.mysql.com/doc/refman/8.0/en/glossary.html#glos_next_key_lock) to block insertions by other sessions into the gaps covered by the range. For information about gap locks and next-key locks, see [Section 15.7.1, “InnoDB Locking”](https://dev.mysql.com/doc/refman/8.0/en/innodb-locking.html).
> 
> I believe we fall into the latter bucket -- does that mean we lock the entire range scanned? In the worst case (exactly 100 deleted messages in the table), it would seem we'd lock the entire table by virtue of the `LIMIT 100` semantics.
> 
> Should I rework this to only delete empty messages explicitly in the page in question, like Postgres does?

@lieut-data Yeah I think you're right, since we don't have an index on just `UserId` or `CreateAt` there's no uniqueness so if we're not limiting the selection by the primary key (ie. the only unique index on `Drafts`), we would probably be locking the whole table.

Unless @agnivade knows something we don't, I would say we do the same thing we do for Postgres where we filter 100 records at a time using a subquery, then join on the primary key.

agnivade commented:
Yep, good point Jesse. Let's shift everything to the postgres model. 

There was also something else that came to mind after you posted the diagram. In your example of - 50 drafts - 50 empty - 50 drafts, let's say we delete the first 50 with a page size of 100. Now our offset is at the end of 100 rows, but then 50 rows from that 100 is now deleted. Although it will work fine because overall, the order still doesn't change, but usually your mental model while paginating thinks of a static data set.

hanzei reviewed (commented):
https://github.com/mattermost/mattermost/pull/24275 changed quite a lot in how logging works for workers. It would be great to update your PR to make use of the new pattern. 

On the other hand, that is not a blocking comment. I'm happy to work on that post-merge,

lieut-data commented:
> There was also something else that came to mind after you posted the diagram. In your example of - 50 drafts - 50 empty - 50 drafts, let's say we delete the first 50 with a page size of 100. Now our offset is at the end of 100 rows, but then 50 rows from that 100 is now deleted. Although it will work fine because overall, the order still doesn't change, but usually your mental model while paginating thinks of a static data set.

Good call out, @agnivade. Thank you for helping us move away from the older paging paradigm of `OFFSET X LIMIT 100` :)

As I'm digging into MySQL, paginating this particular table with the dual columns forming the primary key is exceptionally painful. Without CTEs (looking at you MySQL v5.7!), I'm left either creating a temporary table from which to delete or repeating the query 100 times. (Or maybe there's still some cryptic SQL to be discovered -- suggestions welcome!)

Had there been a `model.NewId()` for the primary key (with a `UNIQUE` constraint on the columns in question), the pagination would have been trivial. I'm wondering if we should require new tables always having single column primary keys?

agnivade commented:
Ahh that's true, no CTEs for 5.7. Woops, this is bad. A temp table sounds fine to me. Soon we will move away from MySQL anyways, so a temporary pain. :)

>  I'm wondering if we should require new tables always having single column primary keys?

Already ahead of you: https://community.mattermost.com/core/pl/di3io5cr4j8cuyetfisuz6u1qr

lieut-data commented:
/update-branch

lieut-data commented:
> Ahh that's true, no CTEs for 5.7. Woops, this is bad. A temp table sounds fine to me. Soon we will move away from MySQL anyways, so a temporary pain. :)

I was able to find the right incantation to get this working with an `INNER JOIN`, now passing the newly added draft store tests. Next week I'll take a stab at the job refactoring.

lieut-data commented:
@agnivade & @devinbinnie, heads up I've completed the previously discussed rework to introduce a more generic `BatchMigrationWorker`, but along the way I wanted to address what I think is a shortcoming in the current logging strategy: https://github.com/mattermost/mattermost/pull/24650.

Once that's merged, I'll update this branch and invite you review :)

lieut-data commented:
@devinbinnie, I can't assign you as a reviewer to your own PR, so I've added you as assignee. Hopefully the proposed changes reflect all of our combined discussions on this front. Let me know if I can improve anything :)

devinbinnie reviewed (commented):
@lieut-data Thanks for taking on that task, I quite like the generic `BatchMigrationWorker`, no comments from my end on the code.

One question from a change management perspective, are we confident that we want this all to go in as one change? Is there anything we could/should do to break it up?

Given how long this PR has been running and how much review time it has had, I doubt it'll get reverted, but something to consider.

Happy to merge as-is as well :)

lieut-data commented:
Thanks, @devinbinnie!

> One question from a change management perspective, are we confident that we want this all to go in as one change? Is there anything we could/should do to break it up?

The biggest change that I wanted to pull out from this PR already got merged as https://github.com/mattermost/mattermost/pull/24650. The only other thing that I think might get pulled out of this PR might be the `main_test.go` and `helper_test.go` changes, but they are only used with the batch migration unit tests. We could theoretically pull out the batch migration worker itself, but without its first corresponding implementation, I'm not certain if that's valuable. Open to other ways of slicing and dicing if I'm missing something!

devinbinnie commented:
> Thanks, @devinbinnie!
> 
> > One question from a change management perspective, are we confident that we want this all to go in as one change? Is there anything we could/should do to break it up?
> 
> The biggest change that I wanted to pull out from this PR already got merged as #24650. The only other thing that I think might get pulled out of this PR might be the `main_test.go` and `helper_test.go` changes, but they are only used with the batch migration unit tests. We could theoretically pull out the batch migration worker itself, but without its first corresponding implementation, I'm not certain if that's valuable. Open to other ways of slicing and dicing if I'm missing something!

Normally I'm in favour of slicing things up as best we can, but since a) this PR has been around long enough :P and b) I think we've done our due diligence to review it meticulously I think we're fine to leave as-is. 1/5.

To be safe, I'll re-request a review from @agnivade before we merge.

agnivade commented on server/channels/jobs/batch_migration_worker.go:
```diff
@@ -0,0 +1,234 @@
+	job.Data = nil
+	job.Progress = 0
+
+	if _, err := worker.store.Job().UpdateOptimistically(job, model.JobStatusInProgress); err != nil {
```
Any reason we aren't directly setting the status to pending here? Seems like we are making yet another call just to set the status to pending.

lieut-data replied:
Good eye! I got wrapped in the subtleties of all the existing `setJob*` methods and missed that I could just set `job.Status = model.JobtStatusPending` directly above. Fixing!

agnivade commented on server/channels/jobs/batch_migration_worker.go:
```diff
@@ -0,0 +1,234 @@
+				return
+			}
+
+			worker.logger.Trace("Worker: processing batch", mlog.Any("data", job.Data))
```
Would suggest to keep this at debug to keep things simple. The expectation is that except the trace settings we have for things like Elasticsearch, LDAP, setting log level to debug will log everything.

lieut-data replied:
Ack, re: keeping things simple. I'm inclined to remove this log altogether given the potential for log spamming, especially for migrations that one day touch millions of rows one day.

agnivade replied:
This will be called once per batch right? 0/5 but this might still be useful in cases where the job just gets randomly stuck. But not a blocker from my side.

agnivade commented on server/channels/jobs/batch_migration_worker.go:
```diff
@@ -0,0 +1,234 @@
+}
+
+// MakeBatchMigrationWorker creates a worker to process the given migration batch function.
+func MakeBatchMigrationWorker(jobServer *JobServer, store store.Store, app BatchMigrationWorkerAppIFace, migrationKey string, timeBetweenBatches time.Duration, doMigrationBatch func(model.StringMap, store.Store) (model.StringMap, bool, error)) model.Worker {
```
Nit: It makes it slightly more readable to name the variables in the `doMigrationBatch` function `func(jobData model.StringMap, store store.Store)`. I had to look at the function definition to see what actually is `model.StringMap`.

agnivade reviewed (approved):
Looks good, thanks!
